tools/gfx/render-graphics-common.cpp

```diff
@@ -859,16 +800,66 @@ class GraphicsCommonShaderObject : public IShaderObject, public RefObject
     {
         if (offset.bindingRangeIndex < 0)
             return SLANG_E_INVALID_ARG;
-        if (offset.bindingRangeIndex >= m_layout->getBindingRangeCount())
+        auto layout = getLayout();
+        if (offset.bindingRangeIndex >= layout->getBindingRangeCount())
             return SLANG_E_INVALID_ARG;
-        auto& bindingRange = m_layout->getBindingRange(offset.bindingRangeIndex);
+        auto& bindingRange = layout->getBindingRange(offset.bindingRangeIndex);
 
         auto& slot = m_combinedTextureSamplers[bindingRange.baseIndex + offset.bindingArrayIndex];
         slot.textureView = textureView;
         slot.sampler = sampler;
         return SLANG_OK;
     }
 
+public:
+    // Appends all types that are used to specialize the element type of this shader object in `args` list.
+    virtual Result collectSpecializationArgs(ExtendedShaderObjectTypeList& args) override
+    {
+        if (!m_bindingFinalized)
+            return SLANG_FAIL;
+        auto& subObjectRanges = getLayout()->getSubObjectRanges();
+        // The following logic is built on the assumption that all fields that involve existential types (and
+        // therefore require specialization) will results in a sub-object range in the type layout.
+        // This allows us to simply scan the sub-object ranges to find out all specialization arguments.
+        for (Index subObjIndex = 0; subObjIndex < subObjectRanges.getCount(); subObjIndex++)
+        {
+            // Retrieve the corresponding binding range of the sub object.
+            auto bindingRange = getLayout()->getBindingRange(subObjectRanges[subObjIndex].bindingRangeIndex);
+            switch (bindingRange.bindingType)
+            {
+            case slang::BindingType::ExistentialValue:
+            {
+                // A binding type of `ExistentialValue` means the sub-object represents a interface-typed field.
+                // In this case the specialization argument for this field is the actual specialized type of the bound
+                // shader object. If the shader object's type is an ordinary type without existential fields, then the
+                // type argument will simply be the ordinary type. But if the sub object's type is itself a specialized
+                // type, we need to make sure to use that type as the specialization argument.
+
+                // TODO: need to implement the case where the field is an array of existential values.
+                SLANG_ASSERT(bindingRange.count == 1);
+                ExtendedShaderObjectType specializedSubObjType;
+                SLANG_RETURN_ON_FAIL(m_objects[subObjIndex]->getSpecializedShaderObjectType(&specializedSubObjType));
```

Comment:
Same issue as for the case in the CUDA back-end. I'd expect to see two levels of nesting here: one over the sub-object ranges and one over the sub-objects in each range. The actual `subObjIndex` should be something computed from the `bindingRange` for the sub-object range plus the index of the particular sub-object in the inner loop.

This is probably fine to leave as-is for now, but it definitely isn't 100% correct.